### DISCUSSION TITLE
Counterbalancing session blocks

### DISCUSSION THREAD

**Discussion by EfiLmp** [immersivecognition/unity-experiment-framework#147] [General]
Hello everyone!

I am new to UXF (and Unity in general) and I am currently building an experiment with 1 practice and 4 experimental blocks. The practice block needs to always be presented first, thought to avoid order effects the experimental blocks need to be randomly presented per participant. At the moment I struggle to find how to shuffle only the experimental session blocks. For example session.blocks.Shuffle(); is randomizing all the session blocks, including practice block. 

I apologize in advance if this is already answered, but it has not come to my attentions through my search to other threads. Any thoughts or guidance on how to acheave this, would be greatly appreciated.

Thank you in advance for your time.

**Comment by jackbrookes** [comment]
Hi! This is a good question. There are a few ways you could do this: 

1. You could shuffle the blocks as you are doing, then force the practice block to be at the start, by removing it and inserting it again:

```cs
session.blocks.Shuffle();
session.blocks.Remove(practiceBlock);
session.blocks.Insert(0, practiceBlock);
```

2. You could re-order the blocks using something like this to force the practiceBlock to the start:

```cs
using System.Linq;

...
practiceBlock.settings.SetValue("practice", true);

session.blocks = session.blocks
    .OrderBy(block => block.settings.GetBool("practice", false) ? -1 : Random.value)
    .ToList();
```

This orders by a random value between `0-1`, except the block that has a setting `"practice"` set to `true`, in which case it has the value `-1` which puts it at the start of the list.

**Reply by EfiLmp** [reply]
Thank you so much for your reply and for your time! It works!
